grails-app/services/au/org/emii/aatams/detection/DetectionExtractService.groovy

```diff
@@ -59,39 +57,48 @@ class DetectionExtractService extends AbstractStreamingExporterService {
     protected void writeCsvData(final filterParams, OutputStream out)
     {
         filterParams.sql = new Sql(dataSource)
+        filterParams.sql.fetchSize = getFetchSize()
+        filterParams.sql.autoCommit = false
         filterParams.projectPermissionCache = [:]
 
         super.writeCsvData(filterParams, out)
     }
 
-    protected def readData(filterParams)
+    protected def eachRow(filterParams, closure)
     {
-        return extractPage(filterParams)
+        def startTime = System.currentTimeMillis()
+        def query =  new QueryBuilder().constructQuery(filterParams)
+        def releaseIsProtectedCache = [:]
+        int count = 0
+        
+        filterParams.sql.eachRow(query.getSQL(), query.getBindValues()) { row ->
+            def sanitisedRows = applyEmbargo([row.toRowResult()], filterParams, releaseIsProtectedCache)
+            sanitisedRows.each { sanitisedRow ->
+                closure.call(sanitisedRow)
+            }
+            count++
+        }
```

Comment:
Yeah, something like that might be better.

That snippet sort of doubles-up some of the logic in the DetectionVisibilityChecker class. But I think this would be essentially the same thing:

    def checker = new DetectionVisibilityChecker(row, params, permissionUtilsService, releaseIsProtectedCache)
    def checkedRow = checker.apply()
    if (checkedRow) {
        closure.call(checkedRow)
    }